src/content/app-architecture/case-study/ui-layer.md

```diff
@@ -0,0 +1,652 @@
+
+```dart title=home_viewmodel.dart
+class HomeViewModel {
+  HomeViewModel({
+    required BookingRepository bookingRepository, 
+    required UserRepository userRepository,
+  }) : 
+    // Repositories are manually assigned because they're private members
+    _bookingRepository = bookingRepository, 
+    _userRepository = userRepository;
+
+  final BookingRepository _bookingRepository;
+  final UserRepository _userRepository;
+  // ... 
+}
+```
+
+ViewModels are always dependent on data repositories, 
+which are provided as arguments to the ViewModel's constructor. 
+ViewModels and repositories have a many-to-many relationship, 
+and most ViewModels will depend on multiple repositories.
+
+As in the example code above, 
+repositories should be private members on the ViewModel, 
+otherwise Views would have direct access to 
+the data layer of the application.
+
+### UI state
+
+The output of a ViewModel is data that a view needs to render, generally
+referred to as **UI State**, or just state. UI state is an immutable snapshot of
+data that is required to fully render a view.
+
+![A screenshot of the booking screen of the compass app.](/assets/images/docs/app-architecture/case-study/mvvm-case-study-ui-state-highlighted.png)
+
+The ViewModel exposes state as public members. 
+On the ViewModel in the following code example,
+the exposed data is a `User` object, 
+as well as the user's saved itineraries which 
+are exposed as an object of type `List<TripSummary>`.
+
+```dart title=home_viewmodel.dart
+class HomeViewModel {
+  HomeViewModel({
+   required BookingRepository bookingRepository, 
+   required UserRepository userRepository,
+  }) : _bookingRepository = bookingRepository,
+      _userRepository = userRepository;
+  
+  final BookingRepository _bookingRepository;
+  final UserRepository _userRepository;
+
+  User? _user;
+  User? get user => _user;
+
+  List<BookingSummary> _bookings = [];
+  UnmodifiableListView<BookingSummary> get bookings => UnmodifiableListView(_bookings);
+
+// ... 
+}
+```
+
+As mentioned, the UI state should be immutable.
+This is a crucial part of bug-free software.
+
+The compass app uses the [freezed][] package to 
+enforce immutability on data classes. For example, 
+the following code shows the `User` class definition. 
+`freezed` provides deep immutability, 
+and generates the code for useful methods like 
+`[copyWith](freezed.copywoith)`and `[toJson](jsonseralizable.toJson)`.
+
+```dart title=user.dart
+@freezed
+class User with _$User {
+  const factory User({
+    /// The user's name.
+    required String name,
+
+    /// The user's picture URL.
+    required String picture,
+  }) = _User;
+
+  factory User.fromJson(Map<String, Object?> json) => _$UserFromJson(json);
+}
+
+```
+
+:::note
+In the ViewModel example, 
+two objects are needed to render the view. 
+As the UI state for any given model grows in complexity, 
+a ViewModel may have many more pieces of data from 
```

Comment:
["may" should be avoided](https://developers.google.com/style/word-list#may) outside of legal text.

```suggestion
a ViewModel might have many more pieces of data from 
```